### ISSUE TITLE
puppet provisioner does not work with puppet_whitelist_exit_code and retry_on_exit_code

### ISSUE THREAD

**Issue by stefanandres** [neillturner/kitchen-puppet#148]
We need to retry puppet runs which fail for temporary reasons (repos not available and stuff).
For that we need to enable puppet_detailed_exitcodes (if not, puppet will just exit with $? == 0).
If puppet_detailed_exitcodes	 is enabled, we need to retry for $? 4 and 6. If puppet_detailed_exitcodes is enabled though, $? == 2 is the new "succeeed" exit code, so we need to whitelist it with puppet_whitelist_exit_code.

Combing those two option will break things:

````
  puppet_detailed_exitcodes: true
  puppet_whitelist_exit_code: 2
  retry_on_exit_code:
    - 0
    - 4
    - 6
  max_retries: 2
````

kitchen-puppet will run:
```
export MANIFESTDIR='/tmp/kitchen/manifests'; sudo -E puppet apply /tmp/kitchen/manifests/role_admin/site.pp --modulepath=/tmp/kitchen/modules --fileserverconfig=/tmp/kitchen/fileserver.conf     --detailed-exitcodes    ; [ $? -eq 2 ] && exit 0
```

The this command will ALWAYS exit with $? == 1 when $? != 2 because of the last command. This will never trigger the retry.

**Comment by stefanandres** [comment]
As a workaround, we can use 
```
retry_on_exit_code:
  - 1
```

**Comment by neillturner** [comment]
With PR  #146
it generates the command 
```
export MANIFESTDIR='/tmp/kitchen/manifests'; sudo -E puppet apply /tmp/kitchen/manifests/site.pp --modulepath=/tmp/kitchen/modules --fileserverconfig=/tmp/kitchen/fileserver.conf     --detailed-exitcodes -v -d  ; RC=$?; [ $RC -eq 2 ] && exit 0; exit $RC
```
```
export MANIFESTDIR='/tmp/kitchen/manifests';sudo -E puppet apply /tmp/kitchen/manifests/site.pp --modulepath=/tmp/kitchen/modules --fileserverconfig=/tmp/kitchen/fileserver.conf     --detailed-exitcodes   ; [ $? -eq 2 ] && exit 0
```
so this should fix this too i think
closing fixed in version 1.45.3
